asf_heat_pump_suitability/getters/get_dno_datasets.py

```diff
@@ -0,0 +1,467 @@
+
+
+def generate_enw_gdf() -> gpd.GeoDataFrame:
+    """
+    Generate a GeoDataFrame for Electricity North West (ENW) primary substations and their service areas.
+
+    The function processes substation data from ENW's Distribution Future Energy Scenarios (DFES)
+    and Network Development Plan (NDP) datasets. It combines point locations of substations with
+    their corresponding service area polygons (Voronoi polygons) and capacity/demand information.
+
+    Returns:
+        gpd.GeoDataFrame: DataFrame with EPSG:4326 (WGS84) projection containing:
+            - id: Unique substation identifier
+            - firm_capacity_mva: Maximum power the substation can safely deliver (MVA)
+            - peak_demand_mva: Maximum observed power demand at the substation (MVA)
+            - geo_shape: Polygon geometry representing the substation's service area
+            - operator: Distribution network operator code ("ENW")
+    """
+    enw_demand = base_getters.get_df_from_parquet_s3_path(
+        config["data_source"]["ENW_dfes_primaries"]
+    ).to_pandas()
+    enw_substations = base_getters.get_df_from_parquet_s3_path(
+        config["data_source"]["ENW_ndp_headroom"]
+    ).to_pandas()
+    enw_shape = base_getters.get_df_from_parquet_s3_path(
+        config["data_source"]["ENW_ndp_voronoi"]
+    ).to_pandas()
+
+    # Filter data for primary substations, current year, and best view scenario
+    enw_substations = enw_substations[
+        (enw_substations["substation_type"] == "PRIMARY")
+        & (enw_substations["year"] == "2024")
+        & (enw_substations["scenario"] == "1 - BEST VIEW")
+        & (enw_substations["status"] == "FIRM")
+    ]
+    enw_demand = enw_demand[
+        (enw_demand.group == "PRIMARY")
+        & (enw_demand.year == "2024")
+        & (enw_demand.scenario == "Best View")
+    ]
+
+    # Merge substation and demand data
+    enw = pd.merge(
+        enw_substations[["substation", "headroom_mva", "geopoint"]],
+        enw_demand[["number", "maximum_demand_mva_per_primary_substation"]],
+        left_on="substation",
+        right_on="number",
+    )
+
+    # Calculate firm capacity and add operator information
+    enw["firm_capacity_mva"] = (
+        enw["maximum_demand_mva_per_primary_substation"] + enw["headroom_mva"]
+    )
+    enw["operator"] = "ENW"
+    enw["geopoint"] = enw["geopoint"].apply(parse_binary_geometry)
+
+    # Create GeoDataFrame and perform spatial join with shape data
+    enw_gdf = gpd.GeoDataFrame(enw, geometry="geopoint", crs=CRS)
+    # polygon geometry of substation distribution area
+    enw_shape["geo_shape"] = enw_shape["geo_shape"].apply(parse_binary_geometry)
+    # point geometry of substation location
+    enw_shape["geo_point_2d"] = enw_shape["geo_point_2d"].apply(parse_binary_geometry)
+    enw_shape_gdf = gpd.GeoDataFrame(enw_shape, geometry="geo_shape", crs=CRS)
+    enw_df = gpd.sjoin(enw_gdf, enw_shape_gdf, how="right", predicate="intersects")
```

Comment:
The point geometry should be close to the centre of the distribution area since the dataframe containing the distribution areas has a row per substation. I've tested with both joins and confirmed that there isn't any difference in the resulting dataframes so we're free of these edge cases.

Reply:
Thanks for checking with the different joins! Sounds good.